### ISSUE TITLE
High memory usage

### ISSUE THREAD

**Issue by MartinDemberger** [Azure/azure-cosmos-dotnet-v3#4027]
**Describe the bug**
We have a service running on two nodes. After a load test I see a high rise of memory usage. I took a memory dump of both and see most of memory is consumed by
* `Task+ContingentProperties` (about 23%)
* `Task<Object>` (about 21%)
* `PooledTimer` (about 19%)

If I look at Retained Size I see a usage of 84% on TimerPool (502 instances).

My application is reading and writing to Azure CosmosDB in a highly parallelized way (several Parallel.ForEachAsync).

**To Reproduce**
Sorry, but I can't reproduce this outside of the test environment.

**Expected behavior**
Memory consumption is lower.

**Actual behavior**
Server memory is eaten up somewhere.

**Environment summary**
SDK Version: 3.35.2 (Microsoft.Azure.Cosmos)
OS Version (e.g. Windows, Linux, MacOSX): Windows Server 2019 Datacenter

**Additional context**
I guess this is memory is consumed by the connection pool for cosmos db. I have configured it as
```csharp
MaxTcpConnectionsPerEndpoint = 1000, // Default: 65535
IdleTcpConnectionTimeout = TimeSpan.FromHours(2), // Default: unlimited
```
to limit the memory usage.

Sorry if this is the wrong place for this question but I don't know where to ask elsewhere.

**Comment by ealsur** [comment]
`PooledTimer` is indeed related to the [TCP connections](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/sdk-connection-modes#volume-of-connections) but the size of the memory should not be high regardless of the number of objects.

`ContingentProperties` is not part of this SDK and `Object` is not clear where it is coming from.

* What is the actual memory size (not percentages or object counts) you are seeing for PooledTimer?
* Are you following the Singleton pattern? Can you share Diagnostics obtained from operations within the environment?
* What is the core/memory available in the instance?

**Comment by MartinDemberger** [comment]
Thank you very much for your help. I will try to answer

* I see a direct memory size of 169.54 MB by `PooledTimer`.
* You mean Singleton pattern for `DocumentClient`? Yes I do only create 2 instances (one for bulk operations and one for normal). But in the dump I can see 333 instances so I think they come from one of the external libraries I use to execute Gremlin-Queries.
* The node has only 8 GB RAM and a lot of other processes running.

I will try to find the creation of `DocumentClient`. I update this issue when I have further infos.

**Comment by MartinDemberger** [comment]
I didn't find the creation of `DocumentClient` by now but I found that the foreign library also uses a singleton pattern (there are exactly 16 instances of `GremlinClient`). This is my fault and I will reduce them to one.

On a local test I was able to produce a similar situation. Here you can see the memory consumption:
![image](https://github.com/Azure/azure-cosmos-dotnet-v3/assets/3066149/2467fe3f-5634-4dba-85fb-f2110701f499)

Can you give me a hint what I can analyse any further?

I will try to reduce the parallelization by replacing all `Parallel.ForEachAsync` with a real foreach-loop. This wouldn't be a accepable solution but maybe this can lead to a better understanding.

I forgot to mention that I use several `ChangeFeedProcessor` inside the process.

**Comment by ealsur** [comment]
The number of TimerPool/PooledTimer is related to the number of connections which is related to the number of clients.

I don't know what `GremlinClient` is, but if this is a wrapper around the C# Cosmos DB SDK, then it will also generate instances of these types.

The more connections, the more timers are used.

Looking at the screenshot, there are only 34MB on this, is that an issue?

**Comment by MartinDemberger** [comment]
You are right, GremlinClient is a wrapper as far as I know.
The screenshot is from my local test. Here is the one from the server dump:
![image](https://github.com/Azure/azure-cosmos-dotnet-v3/assets/3066149/621afd74-8629-4880-8d6c-7a48ff035db8)

I find it alarming that most of the memory is used by Task metadata and by PooledTimer.
On the server there are about 333 instances of `StoreClient`/`DocumentClient`, 334 instances of `HttpClient` which is more than I expected (but this may be ok).
This 333 clients create 670 `TimerPool` s which holds total of 2,777,678 `PooledTimer`s. Each one with it's own instance of `Task<object>` and `Task+ContingentProperties`.

I'm desperate because I don't find the positions where this Tasks are created. On the other side I am supprised that there are so many heavy weight objects created inside the SDK (or the framework).

I think the root cause is in my code but I don't find any way to identify it.
Is there a known problem with a lot of parallel executed queries? Is a ChangeFeed with a lot of changes known to consume a lot of memory?

**Comment by ealsur** [comment]
> On the server there are about 333 instances of StoreClient/DocumentClient

This is the problem, you have 333 client instances. Each client instance has an independent client pool, that is why it's **critical** to follow the Singleton pattern as our guidelines say.

The large size in PooledTimer objects is due probably to the large size in client instances.

Where are these instances created? Really only you can tell.

**Comment by MartinDemberger** [comment]
I found my error. Now I only have 2 DocumentClients. The memory usage is reduced but it's still to high.

![image](https://github.com/Azure/azure-cosmos-dotnet-v3/assets/3066149/3ef256ac-542a-47a3-bc3b-8065468f26a2)

I will try further to find the source where this Tasks are created:
![image](https://github.com/Azure/azure-cosmos-dotnet-v3/assets/3066149/3bc69c92-e790-4d48-b876-81a52df0902f)

**Comment by MartinDemberger** [comment]
I was able to reproduce this behaviour in a simple console application https://github.com/MartinDemberger/CosmosDbRam
It's a small test so the numbers are much below the above but as you can see there is a lot of memory consumed by objects which I think should only be short living.
In my real world application I have a lot of users warking in parallel. I have several ChangeFeeds attached which trigger index updates. So there is a lot of parallel access happening.

Can you give me an hint what is wrong in my code and how I can reduce the memory consumption?

btw: I only have one Client but there are 2 TimePool instances.

**Comment by ealsur** [comment]
@MartinDemberger On the screenshots you provided, there are only 9MB used on PooledTimer, is this considered high to you? There is more memory being taken by other sources, why is this critical?

As I mentioned before, the volume of PooledTimer is dependent on the number of TCP connections, it has nothing to do with the users of the application.

The number of TCP connections is dependent on the partitions in the container, this is all described in our docs: https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/sdk-connection-modes#volume-of-connections

I checked the console application, it has nothing wrong.

If the environment is one that does not meet the requirements for Direct mode (either the number of cores is lower than 4 or has limits in the number of TCP connections) consider user Gateway mode.

**Comment by ealsur** [comment]
> I understand that PooledTimers depends on TCP connections. But more users leads to more doncurrent requests and this means more connections as the documentation says.

That's not completely accurate. As per https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/sdk-connection-modes#volume-of-connections, a new connection will be created only when the volume of concurrent requests hits the configured limit. You can change this limit if desired.

As far as idle connections, on the same link, it says how to defined the settings to remove idle connections.

By default, connections are never closed, to be reused in subsequent operations.

>  Our test environment from the initial report is currently blocked.

Blocked due to memory usage?

> But in a production near environment I had other numbers (see the initial numbers).

These initial numbers were with the problem that the code was creating 333 clients though?

> If you take in count that this is a statless microservice which should use a total of about 100 - 300 MB

Are there other limitations for this environment? How many Cores are assigned to the microservice? What's the CPU?

**Comment by ealsur** [comment]
> MaxRetryWaitTimeOnRateLimitedRequests = TimeSpan.FromHours(2),
> RequestTimeout = TimeSpan.FromHours(2),

MaxRetryWaitTimeOnRateLimitedRequests has nothing to do, it is the setting to retry on 429s.
2 Hours for RequestTimeout is really pointless. RequestTimeout is meant to govern the max latency a TCP request can have on the wire. Normal values go from 1-60 seconds. This is supposed to mean "My application will only consider acceptable waiting for a network request up to X time, after which, consider it a timeout", I doubt your application can wait 2 hours for a single network request? If so, then there are no users waiting for that response? https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/troubleshoot-dotnet-sdk-request-timeout?tabs=cpu-new#customize-the-timeout-on-the-azure-cosmos-db-net-sdk

> The run on a Azure VM Standard_B2ms together with other services. They are implemented as a Azure Service Fabric Stateless Service.



>  is there any possibility to get the source and create pull requests for TimerPool and PooledTimer?

Source is: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/PooledTimer.cs

**Comment by MartinDemberger** [comment]
> > MaxRetryWaitTimeOnRateLimitedRequests = TimeSpan.FromHours(2),
> > RequestTimeout = TimeSpan.FromHours(2),
> 
> MaxRetryWaitTimeOnRateLimitedRequests has nothing to do, it is the setting to retry on 429s. 2 Hours for RequestTimeout is really pointless. RequestTimeout is meant to govern the max latency a TCP request can have on the wire. Normal values go from 1-60 seconds. This is supposed to mean "My application will only consider acceptable waiting for a network request up to X time, after which, consider it a timeout", I doubt your application can wait 2 hours for a single network request? If so, then there are no users waiting for that response? https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/troubleshoot-dotnet-sdk-request-timeout?tabs=cpu-new#customize-the-timeout-on-the-azure-cosmos-db-net-sdk

My application processes a lot of imports and exports. So the user waits for it (can't be done by night) but dependent of the size this can last for minutes or even hours. So the UI informs the user when the import or export is done.

When there are synchron UI actions I use the `CancellationToken` to stop the request.

I will reduce the timeout to 30 minutes. This might reduce the RAM usage after a burst to 30 Minutes instead of 2 hours. It's not my prefered change but should be acceptable.

I fear this reduces the problem but doesn't solve the problem when a lot of actions are running at the same time.

> > The run on a Azure VM Standard_B2ms together with other services. They are implemented as a Azure Service Fabric Stateless Service.
> 
> Standard_B2ms has 2 cores. Based on https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/best-practice-dotnet, you probably should be using Gateway mode.
> 
> > is there any possibility to get the source and create pull requests for TimerPool and PooledTimer?
> 
> Source is: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/PooledTimer.cs

Thanks, I will have a look at it.

**Comment by MartinDemberger** [comment]
I found a simple change in `PooledTimer` and `TimerPool` which would make the situation a lot better. See the new PR.

**Comment by ealsur** [comment]
> My application processes a lot of imports and exports. So the user waits for it (can't be done by night) but dependent of the size this can last for minutes or even hours. So the UI informs the user when the import or export is done.

Request Timeout is unrelated to the volume of requests. It is a per-network interaction timeout. 30 minutes is still a wildly large value. 
Your user experience time and the network-timeout are 2 separate things. A user might be ok with waiting 30 minutes for their data to be processed, but this is not what the Request Timeout enforces. Network requests do not remain in the wire for 30 minutes.

**Have you tried using a normal timeout value, like 10 seconds or leaving the default?**

> I fear this reduces the problem but doesn't solve the problem when a lot of actions are running at the same time.

Number of actions again is not related to the timeout. Actions might translate to number of network requests.

Regarding the machine SKU comment, have you tried using Gateway mode?

**Comment by MartinDemberger** [comment]
> > My application processes a lot of imports and exports. So the user waits for it (can't be done by night) but dependent of the size this can last for minutes or even hours. So the UI informs the user when the import or export is done.
> 
> Request Timeout is unrelated to the volume of requests. It is a per-network interaction timeout. 30 minutes is still a wildly large value. Your user experience time and the network-timeout are 2 separate things. A user might be ok with waiting 30 minutes for their data to be processed, but this is not what the Request Timeout enforces. Network requests do not remain in the wire for 30 minutes.
> 
> **Have you tried using a normal timeout value, like 10 seconds or leaving the default?**

Sorry to ask again:
If I have a timeout of 30 minutes and the request on the wire is completed after a few milliseconds this shouldn't be a problem. If I change the timeout to 6 seconds and the request is completed after a few milliseconds it's (beside the memory usage) completely the same situation. Or is there something I doesn't see?

I have this high timeout because a timeout is a worst case scenario. Depending on the executed request that timeouts the result is a retry of the request, a undo of previous actions (a lot of deletes) with a retry or a failed import which will also results in a retry of the complete import triggered by the user.
Because of this I have a very high timeout to reduce the risk it appears. My business code uses `CancellationToken` to signal situation where the actions should and can be stopped.

I have seen complex queries which are on the wire for more then a few seconds. So with my situation in mind I configured this high value.
Are there other timeouts which will occure?

> > I fear this reduces the problem but doesn't solve the problem when a lot of actions are running at the same time.
> 
> Number of actions again is not related to the timeout. Actions might translate to number of network requests.

You say that reducing the timeout doesn't solve my problem?

> Regarding the machine SKU comment, have you tried using Gateway mode?

Currently I use the default value, I will set the Gateway mode explicit on my next test.
As I understand this is almost identical to Direct mode. The difference is that direct mode only uses host resolution if connection is broken and Gateway *always* does the resolution.
How will this affect my memory problem? Are there less connections used in Gateway mode?
In https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/sdk-connection-modes#available-connectivity-modes are conditions when to use which mode. Because of them I guess Direct mode is the better matching.

**Comment by ealsur** [comment]
> Because of this I have a very high timeout to reduce the risk it appears. My business code uses CancellationToken to signal situation where the actions should and can be stopped.

CT and RequesTimeout cover different areas: https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/troubleshoot-dotnet-sdk-request-timeout?tabs=cpu-new#customize-the-timeout-on-the-azure-cosmos-db-net-sdk

> I have seen complex queries which are on the wire for more then a few seconds. 

Queries have no latency SLA and depending on the complexity of the query can run for up to 5 seconds: https://learn.microsoft.com/en-us/azure/cosmos-db/concepts-limits#per-request-limits

> As I understand this is almost identical to Direct mode.

These are **very** different Transports. Direct mode as explained in the link you shared, connects to the Backend Replicas directly using TCP protocol, the volume of connections is higher than Gateway mode, which just uses the HTTP protocol. The handling of these TCP connections is more resource intensive, versus Gateway mode which just uses the .NET HttpClient. Thus the recommendation for Direct mode in the .NET best practices is to have at least 4 cores (https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/best-practice-dotnet).  The link you shared shows the reasons (not conditions really) for which you might want to choose one or the other, but the requirements in terms of available resources are not listed there. Yes, Direct mode gives better latency and performance but it is also consumes more resources. If your scenario is small microservices which run on 2-core machines but should not even use all the machine's available resources but the least possible ones, then Direct mode is not the right fit.

**Comment by MartinDemberger** [comment]
> Queries have no latency SLA and depending on the complexity of the query can run for up to 5 seconds: https://learn.microsoft.com/en-us/azure/cosmos-db/concepts-limits#per-request-limits

Ah, I forgot this quota. Thank you for pointing on it.
With this in mind my high values are absolutely nonsense. I will set my timeouts to 10 sec.

In https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.cosmosclientoptions.requesttimeout?view=azure-dotnet the timeout has default value of 1 minute. I think this value is wrong because in code I see a default of 6 secs.
Also it might help if a hint is added that high values increase the needed resources but doesn't help because of the quota.

> Direct mode gives better latency and performance but it is also consumes more resources. If your scenario is small microservices which run on 2-core machines but should not even use all the machine's available resources but the least possible ones, then Direct mode is not the right fit.

I got the point.

In my tests with the high timeout most of the RAM was consumed by the `TimerPool`. The connections doesn't show significant RAM usage (but was visible). Hopefully the lower timeout will remove the `TimerPool` problem and then I will analyse if the resources of the connections are acceptable. If not I will switch to Gateway Mode.


Thanks for your patience.

**Comment by ealsur** [comment]
> In https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.cosmosclientoptions.requesttimeout?view=azure-dotnet the timeout has default value of 1 minute. I think this value is wrong because in code I see a default of 6 secs.

Yeah, we noticed that. We don't update those public documentation pages and they should be auto-updated with library releases. But something isn't working (I'm following up on that).

`TimerPool` usage in Gateway Mode is minimal/non-existent (as far as I can remember, it is mainly used for TCP Connection/Request monitoring).
